### DISCUSSION TITLE
es/guides/object-counting/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8341] [Docs]
# es/guides/object-counting/

Recuento de objetos mediante Ultralytics YOLOv8

https://docs.ultralytics.com/es/guides/object-counting/

**Comment by valentinrc** [comment]
Hello,

Is it possible to extract the OutCount in a list, dataframe, etc?

I would like to see the final number of the OutCount without downloading the out video.

Thanks.

**Reply by pderrenger** [reply]
@valentinrc hello!

Absolutely, you can access the OutCount without needing to download the output video. After running the object counter, you can extract the count information directly from the `ObjectCounter` instance. Here's a quick example of how you might do this:

```python
from ultralytics import YOLO, ObjectCounter

# Load your model and video
model = YOLO("yolov8n.pt")
cap = cv2.VideoCapture("path/to/video.mp4")

# Initialize the object counter
counter = ObjectCounter()

# Process the video
while True:
    success, frame = cap.read()
    if not success:
        break
    tracks = model.track(frame, persist=True)
    counter.update(tracks)

# Access the final counts
in_count = counter.in_count
out_count = counter.out_count

# For example, print the out count
print("Out Count:", out_count)
```

This will print the final OutCount to the console. You can also store it in a list or a DataFrame as per your requirement. If you need further assistance, feel free to ask. Happy coding! 😊

**Comment by wisrovi** [comment]
Hello,

Is it possible to extract the OutCount and InCount for every class?

for example, a model predict person and car, the model print the number of cars and persons

I try with:
# Access the final counts
in_count = counter.in_counts
out_count = counter.out_counts
but only I finded the counter have a global, joined persons and cars in the same count

I would like to see the final number of the every class without downloading the out video.

Thanks.

**Reply by pderrenger** [reply]
@wisrovi ¡Hola!

Para obtener el recuento de cada clase específica (como personas y autos) sin mezclar, puedes modificar el código para iterar sobre los `tracks` y contar las clases individualmente. Aquí tienes un ejemplo de cómo podrías hacerlo:

```python
from collections import defaultdict

# Inicializa contadores para cada clase
class_counts = defaultdict(int)

# Itera sobre cada track y actualiza el contador de la clase correspondiente
for track in tracks:
    class_id = track['class']  # Asume que 'class' es la clave para el ID de clase en tu objeto track
    class_counts[class_id] += 1

# Ahora class_counts contiene el recuento de cada clase
for class_id, count in class_counts.items():
    print(f"Clase {model.names[class_id]}: {count}")
```

Este código asume que tienes un objeto `tracks` que contiene la información de seguimiento, incluido el ID de clase de cada objeto detectado. `model.names` se utiliza para obtener el nombre de la clase a partir del ID de clase.

Espero que esto te ayude a obtener los recuentos individuales por clase que necesitas. ¡Déjame saber si tienes más preguntas! 😊

**Reply by nrj64** [reply]
Hola una consulta, he hecho una aplicación para contar, al estitlo acuicultura, el problema que tengo es que al deployarla en el server dejo de ver la región de conteo. No solo la parte gráfica sino que el conteo se ve alterado. En cambio al correrlo de forma local funciona bien. Ambas maquinas son linux ubuntu 22.04

**Reply by pderrenger** [reply]
¡Hola! 👋 Parece que tu situación podría estar relacionada con la configuración del servidor o la forma en que se está manejando la visualización de las regiones de conteo en diferentes entornos. Aquí hay un par de sugerencias:

1. **Verificación de la Configuración del Servidor**: Asegúrate de que la configuración del servidor es compatible con la visualización gráfica de las regiones de conteo.

2. **Código de Debugging**: Verifica si hay alguna diferencia en las configuraciones o en las bibliotecas entre tu entorno local y el servidor. A veces, las diferencias en las versiones de las bibliotecas pueden causar problemas.

3. **Región de conteo Dinámica**: Si estás utilizando coordenadas fijas para definir la región de conteo, asegúrate de que estas coordenadas sean válidas dentro de la resolución de video en el servidor.

Para el contado diferenciado por clases, te recomendaría iterar sobre los resultados para filtrar y contar por clase específica. Si estás utilizando `model.track()`, los resultados incluirán las clases detectadas, lo que te permite hacer algo como lo siguiente:

```python
from collections import defaultdict

# Inicializa un contador para cada clase
counts = defaultdict(int)

# Suponiendo que 'tracks' es la lista de tus objetos rastreados
for track in tracks:
    clase = track['class']  # Asegúrate de que este es el campo correcto para tu estructura
    counts[clase] += 1

# Ahora 'counts' tiene el conteo separado por clase
for clase, count in counts.items():
    print(f"Clase {clase}: {count}")
```

Recuerda adaptar el campo `'class'` según la estructura de tu resultado. Espero que esto solucione tu problema. ¡Déjanos saber si necesitas más ayuda!

**Comment by manunava11** [comment]
Hi, any way to hide the line?
Tried line_thickness=0 but it doesnt work.

Thanks!

**Reply by pderrenger** [reply]
Hi there!

Thank you for reaching out! To hide the line, you can set the `line_thickness` parameter to a very small value, such as `0.1`, instead of `0`. This should effectively make the line invisible.

Here's an example of how you can do this:

```python
import cv2
from ultralytics import YOLO, solutions

model = YOLO("yolov8n.pt")
cap = cv2.VideoCapture("path/to/video/file.mp4")
assert cap.isOpened(), "Error reading video file"
w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

# Define region points
region_points = [(20, 400), (1080, 404), (1080, 360), (20, 360)]

# Video writer
video_writer = cv2.VideoWriter("object_counting_output.avi", cv2.VideoWriter_fourcc(*"mp4v"), fps, (w, h))

# Init Object Counter
counter = solutions.ObjectCounter(
    view_img=True,
    reg_pts=region_points,
    classes_names=model.names,
    draw_tracks=True,
    line_thickness=0.1,  # Set to a very small value
)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        print("Video frame is empty or video processing has been successfully completed.")
        break
    tracks = model.track(im0, persist=True, show=False)

    im0 = counter.start_counting(im0, tracks)
    video_writer.write(im0)

cap.release()
video_writer.release()
cv2.destroyAllWindows()
```

If you continue to experience issues, please ensure you are using the latest versions of `torch` and `ultralytics`. You can update them using:

```bash
pip install --upgrade torch ultralytics
```

For more detailed guidance on object counting, you can refer to our [Object Counting Guide](https://docs.ultralytics.com/guides/object-counting/).

Feel free to reach out if you have any more questions or need further assistance! 😊

**Reply by manunava11** [reply]
Hi, i get an error when i set line_thickness = 0.1
Argument 'thickness' is required to be an integer.

**Reply by pderrenger** [reply]
@manunava11 hi there!

Thank you for bringing this to our attention. The `line_thickness` parameter indeed requires an integer value, which is why setting it to `0.1` results in an error. To effectively hide the line, you can set the `line_thickness` to `0`, which should work as intended. If this approach doesn't hide the line, you can alternatively set the `line_thickness` to a very small integer value, such as `1`, and set the `line_color` to match the background color, effectively making it invisible.

Here's an example of how you can do this:

```python
import cv2
from ultralytics import YOLO, solutions

model = YOLO("yolov8n.pt")
cap = cv2.VideoCapture("path/to/video/file.mp4")
assert cap.isOpened(), "Error reading video file"
w, h, fps = (int(cap.get(x)) for x in (cv2.CAP_PROP_FRAME_WIDTH, cv2.CAP_PROP_FRAME_HEIGHT, cv2.CAP_PROP_FPS))

# Define region points
region_points = [(20, 400), (1080, 404), (1080, 360), (20, 360)]

# Video writer
video_writer = cv2.VideoWriter("object_counting_output.avi", cv2.VideoWriter_fourcc(*"mp4v"), fps, (w, h))

# Init Object Counter
counter = solutions.ObjectCounter(
    view_img=True,
    reg_pts=region_points,
    classes_names=model.names,
    draw_tracks=True,
    line_thickness=1,  # Set to a very small integer value
    count_reg_color=(0, 0, 0)  # Set to match the background color
)

while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        print("Video frame is empty or video processing has been successfully completed.")
        break
    tracks = model.track(im0, persist=True, show=False)

    im0 = counter.start_counting(im0, tracks)
    video_writer.write(im0)

cap.release()
video_writer.release()
cv2.destroyAllWindows()
```

Please ensure you are using the latest versions of `torch` and `ultralytics`. You can update them using:

```bash
pip install --upgrade torch ultralytics
```

For more detailed guidance on object counting, you can refer to our [Object Counting Guide](https://docs.ultralytics.com/guides/object-counting/).

Feel free to reach out if you have any more questions or need further assistance! 😊

**Reply by manunava11** [reply]
Solved! Thanks.

**Reply by pderrenger** [reply]
@manunava11 hi there!

I'm glad to hear that your issue is resolved! 🎉 If you have any more questions or need further assistance, feel free to ask. 

For anyone else encountering similar issues, remember that setting `line_thickness` to `0` or a very small integer value should effectively hide the line. If you experience any difficulties, please ensure you are using the latest versions of `torch` and `ultralytics`. You can update them using:

```bash
pip install --upgrade torch ultralytics
```

For more detailed guidance on object counting, you can refer to our [Object Counting Guide](https://docs.ultralytics.com/guides/object-counting/).

Thank you for using Ultralytics YOLOv8, and happy coding! 😊

**Comment by greviewer-edt** [comment]
gracias, podrian por favor poner los videos de ejemplo, gracias
